miden-tx/src/tx_benches/utils.rs

```diff
@@ -0,0 +1,267 @@
+use miden_objects::{
+    accounts::{Account, AccountId, AccountStub},
+    assembly::ModuleAst,
+    notes::NoteId,
+    transaction::{ChainMmr, InputNote, InputNotes, TransactionArgs},
+    BlockHeader,
+};
+use miden_tx::{DataStore, DataStoreError, TransactionHost, TransactionInputs};
+use mock::mock::{
+    account::MockAccountType, notes::AssetPreservationStatus, transaction::mock_inputs,
+};
+use vm_processor::{
+    AdviceExtractor, AdviceInjector, AdviceProvider, ExecutionError, Host, HostResponse,
+    ProcessState,
+};
+
+// CONSTANTS
+// ================================================================================================
+
+/// Number of cycles needed to create an empty span whithout changing the stack state.
+const SPAN_CREATION_SHIFT: u32 = 2;
+
+// BENCHMARK HOST
+// ================================================================================================
+
+/// Wrapper around [TransactionHost] used for benchmarking: `tx_progress` allows to track the
+/// progression of the transaction execution.
+pub struct BenchHost<A: AdviceProvider> {
+    host: TransactionHost<A>,
+    tx_progress: TransactionProgress,
+}
+
+impl<A: AdviceProvider> BenchHost<A> {
+    pub fn new(account: AccountStub, adv_provider: A) -> Self {
+        Self {
+            host: TransactionHost::new(account, adv_provider),
+            tx_progress: TransactionProgress::default(),
+        }
+    }
+}
+
+impl<A: AdviceProvider> Host for BenchHost<A> {
+    fn get_advice<S: ProcessState>(
+        &mut self,
+        process: &S,
+        extractor: AdviceExtractor,
+    ) -> Result<HostResponse, ExecutionError> {
+        self.host.get_advice(process, extractor)
+    }
+
+    fn set_advice<S: ProcessState>(
+        &mut self,
+        process: &S,
+        injector: AdviceInjector,
+    ) -> Result<HostResponse, ExecutionError> {
+        self.host.set_advice(process, injector)
+    }
+
+    fn on_trace<S: ProcessState>(
+        &mut self,
+        process: &S,
+        trace_id: u32,
+    ) -> Result<HostResponse, ExecutionError> {
+        let event = TransactionTrace::try_from(trace_id)
+            .map_err(|err| ExecutionError::EventError(err.to_string()))?;
+
+        use TransactionTrace::*;
+        match event {
+            PrologueStart => self.tx_progress.prologue.set_start(process.clk()),
+            PrologueEnd => self.tx_progress.prologue.set_end(process.clk()),
+            NotesProcessingStart => self.tx_progress.notes_processing.set_start(process.clk()),
+            NotesProcessingEnd => self.tx_progress.notes_processing.set_end(process.clk()),
+            NoteConsumingStart => {
+                self.tx_progress.note_consuming.push(CycleInterval::new(process.clk()))
+            },
+            NoteConsumingEnd => {
+                if let Some(interval) = self.tx_progress.note_consuming.last_mut() {
+                    interval.set_end(process.clk())
+                }
+            },
```

Comment:
I would rename `NoteConsumingStart`/`NoteConsumingEnd` into `NoteExecutionStart`/`NoteExecutionEnd`.

Also, it would be cool if we could capture `NoteId` of the note which is being executed (we should be able to get it from the `process`).

Reply:
I didn't manage to get `NoteId`s from the `process` but I found how to get them from the `Host`, although it cam be done only if corresponding `emit.NOTE_CREATED` event is used, so it works for default notes (for two out of three), but not for P2ID note. Is there a way to get `NoteId` consistently?

Reply:
In kernel memory we keep track of the note that is currently being executed (see [here](https://github.com/0xPolygonMiden/miden-base/blob/main/miden-lib/src/transaction/memory.rs#L31)). You should be able to get this value from memory and then read note ID from the corresponding address within the note (see [here](https://github.com/0xPolygonMiden/miden-base/blob/main/miden-lib/src/transaction/memory.rs#L181)).